### ISSUE TITLE
Unintended retention of old MPI values if mbedtls_mpi_fill_random fails

### ISSUE THREAD

**Issue by guidovranken** [ARMmbed/mbedtls#740]
There are 5 places in the library where ```mbedtls_mpi_fill_random``` is called, but its return value isn't checked:

- ```mbedtls_dhm_make_params```
- ```mbedtls_dhm_make_public```
- ```dhm_update_blinding```
- ```ecp_randomize_jac```
- ```ecp_randomize_mxz```

The function itself is defined as follows:
```c
1866 int mbedtls_mpi_fill_random( mbedtls_mpi *X, size_t size,                                                                                    
1867                      int (*f_rng)(void *, unsigned char *, size_t),
1868                      void *p_rng )
1869 {
1870     int ret;         
1871     unsigned char buf[MBEDTLS_MPI_MAX_SIZE];
1872 
1873     if( size > MBEDTLS_MPI_MAX_SIZE )                       
1874         return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );       
1875                                                             
1876     MBEDTLS_MPI_CHK( f_rng( p_rng, buf, size ) );   
1877     MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( X, buf, size ) );
1878                                                     
1879 cleanup:
1880     return( ret );
1881 }   
```

```f_rng``` can fail due to entropy shortage, memory shortage, or for any other reason if a custom RNG implementation is used for ```f_rng```.

```mbedtls_mpi_read_binary``` can fail due to memory shortage:

```c
 667 int mbedtls_mpi_read_binary( mbedtls_mpi *X, const unsigned char *buf, size_t buflen )
 668 {
 669     int ret;
 670     size_t i, j, n;
 671 
 672     for( n = 0; n < buflen; n++ )
 673         if( buf[n] != 0 )
 674             break;
 675 
 676     MBEDTLS_MPI_CHK( mbedtls_mpi_grow( X, CHARS_TO_LIMBS( buflen - n ) ) );
 677     MBEDTLS_MPI_CHK( mbedtls_mpi_lset( X, 0 ) );
 678 
 679     for( i = buflen, j = 0; i > n; i--, j++ )
 680         X->p[j / ciL] |= ((mbedtls_mpi_uint) buf[i - 1]) << ((j % ciL) << 3);
 681 
 682 cleanup:
 683 
 684     return( ret );
 685 }
```

Every failure path leaves the ```X``` MPI untouched.

For ```ecp_randomize_jac``` and ```ecp_randomize_mxz``` failure in ```mbedtls_mpi_fill_random``` is not important, because the MPI variable they operate on ("```l```") is initialized prior to the loop in which ```mbedtls_mpi_fill_random``` is called, and the loop condition guarantees that ```l``` is set to a non-zero value at some point:

```ecp_randomize_jac```:
```c
1085 
1086     mbedtls_mpi_init( &l ); mbedtls_mpi_init( &ll );
1087 
1088     /* Generate l such that 1 < l < p */
1089     do
1090     {
1091         mbedtls_mpi_fill_random( &l, p_size, f_rng, p_rng );                                                                                 
1092 
1093         while( mbedtls_mpi_cmp_mpi( &l, &grp->P ) >= 0 )
1094             MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &l, 1 ) );
1095 
1096         if( count++ > 10 )
1097             return( MBEDTLS_ERR_ECP_RANDOM_FAILED );
1098     }
1099     while( mbedtls_mpi_cmp_int( &l, 1 ) <= 0 );
```

```ecp_randomize_mxz```:
```c
1468     mbedtls_mpi_init( &l );
1469 
1470     /* Generate l such that 1 < l < p */
1471     do
1472     {
1473         mbedtls_mpi_fill_random( &l, p_size, f_rng, p_rng );                                                                                 
1474 
1475         while( mbedtls_mpi_cmp_mpi( &l, &grp->P ) >= 0 )
1476             MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &l, 1 ) );
1477 
1478         if( count++ > 10 )
1479             return( MBEDTLS_ERR_ECP_RANDOM_FAILED );
1480     }
1481     while( mbedtls_mpi_cmp_int( &l, 1 ) <= 0 );
```

```mbedtls_dhm_make_params``` and ```mbedtls_dhm_make_public``` both operate on ```ctx->X``` as the MPI where ```ctx``` is a ```mbedtls_dhm_context```. I haven't really traced back all the possible call path to see whether execution can arrive in either of these functions with a ```ctx->X``` that has retained a previous value (and thus leaving that value intact should ```mbedtls_mpi_fill_random``` fail), but in ```dhm_update_blinding``` its obvious from the function that the MPI in question (```ctx->Vi```) is already initialized to a certain value:

```c
304     /*
305      * Ok, we need blinding. Can we re-use existing values?
306      * If yes, just update them by squaring them.
307      */
308     if( mbedtls_mpi_cmp_int( &ctx->Vi, 1 ) != 0 )
309     {
310         MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &ctx->Vi, &ctx->Vi, &ctx->Vi ) );
311         MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &ctx->Vi, &ctx->Vi, &ctx->P ) );
312 
313         MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &ctx->Vf, &ctx->Vf, &ctx->Vf ) );
314         MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &ctx->Vf, &ctx->Vf, &ctx->P ) );
315 
316         return( 0 );
317     }
318 
319     /*
320      * We need to generate blinding values from scratch
321      */
322 
323     /* Vi = random( 2, P-1 ) */
324     count = 0;
325     do
326     {
327         mbedtls_mpi_fill_random( &ctx->Vi, mbedtls_mpi_size( &ctx->P ), f_rng, p_rng );
328 
329         while( mbedtls_mpi_cmp_mpi( &ctx->Vi, &ctx->P ) >= 0 )
330             MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->Vi, 1 ) );
331 
332         if( count++ > 10 )
333             return( MBEDTLS_ERR_MPI_NOT_ACCEPTABLE );
334     }
335     while( mbedtls_mpi_cmp_int( &ctx->Vi, 1 ) <= 0 );
```

Here, a ```mbedtls_mpi_fill_random``` could cause the previous ```ctx->Vi``` to be retained.

So I suggest you put a ```MBEDTLS_MPI_CHK``` around these calls.

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-1151

**Comment by sbutcher-arm** [comment]
Fixed in PR's #749, #762 and #763 which are now merged. This issue can be closed.
